tket/src/PauliGraph/PauliGraph.cpp

```diff
@@ -420,4 +423,26 @@ void PauliGraph::to_graphviz(std::ostream &out) const {
   out << "}";
 }
 
+void PauliGraph::sanity_check() const {
+  for (TopSortIterator it = begin(); it != end(); ++it) {
```

Comment:
What exactly caused the segfault? Since the `begin` is called at the beginning, why is the `TopSortIterator` corrupt?

Reply:
This commit has a `pg.sanity_check();` in the constructor `TopSortIterator(const PauliGraph &pg)`. Since `sanity_check()` calls `begin()`, which then calls the `TopSortIterator` constructor, maybe the segfault was caused by this infinite loop?

Reply:
That's a good point and it could be. I saw this segfault on Windows with no changes to the `PauliGraph` code (see https://github.com/CQCL/tket/actions/runs/3313870277/jobs/5472416269 ), and tried to reproduce locally, but maybe I actually failed.

However, the change in this PR _did_ fix the segfault I was observing on Windows. (See https://github.com/CQCL/tket/actions/runs/3314406374/jobs/5476363405 -- the failure on that run is unrelated and happened after the test run successfully completed).

So I'd still like to merge this even without fully understanding the original problem.

Reply:
I agree. The new code is definitely better.